<a href="https://colab.research.google.com/github/jovanrajic/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [ ]:
import pandas as pd
import numpy as np
import matplotlib as plt
from scipy import stats

In [ ]:
### YOUR WORK HERE

columns = ['symboling','norm_loss','make','fuel','aspiration','doors',
           'bod_style','drv_wheels','eng_loc','wheel_base','length','width',
           'height','curb_weight','engine','cylinders','engine_size',
           'fuel_system','bore','stroke','compression','hp','peak_rpm',
           'city_mpg','hgwy_mpg','price']

df = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data", header = None, names = columns)

df.head()

symboling norm_loss         make fuel  ... peak_rpm city_mpg hgwy_mpg  price
0          3         ?  alfa-romero  gas  ...     5000       21       27  13495
1          3         ?  alfa-romero  gas  ...     5000       21       27  16500
2          1         ?  alfa-romero  gas  ...     5000       19       26  16500
3          2       164         audi  gas  ...     5500       24       30  13950
4          2       164         audi  gas  ...     5500       18       22  17450

[5 rows x 26 columns]

In [ ]:
df = df.replace("?", np.NaN)

df.head()

symboling norm_loss         make fuel  ... peak_rpm city_mpg hgwy_mpg  price
0          3       NaN  alfa-romero  gas  ...     5000       21       27  13495
1          3       NaN  alfa-romero  gas  ...     5000       21       27  16500
2          1       NaN  alfa-romero  gas  ...     5000       19       26  16500
3          2       164         audi  gas  ...     5500       24       30  13950
4          2       164         audi  gas  ...     5500       18       22  17450

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [ ]:
df["hgwy_mpg"].mean()

30.75121951219512

In [ ]:
### YOUR WORK HERE

sample1 = df.sample(10, random_state = 30)

In [ ]:
sample1.head()

symboling norm_loss           make  ... city_mpg hgwy_mpg  price
121          1       154       plymouth  ...       31       38   6692
176         -1        65         toyota  ...       27       32  10898
24           1       148          dodge  ...       31       38   6229
88          -1       137     mitsubishi  ...       23       30   9279
72           3       142  mercedes-benz  ...       16       18  35056

[5 rows x 26 columns]

In [ ]:
sample1["hgwy_mpg"].mean()

32.8

In [ ]:
stats.ttest_1samp(sample1["hgwy_mpg"], 30.75121951219512 )

Ttest_1sampResult(statistic=0.8912422391619923, pvalue=0.39600040771093326)

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [ ]:
### YOUR WORK HERE
mpg1 = df["hgwy_mpg"]
mpg2 = sample1["hgwy_mpg"]

In [ ]:
mpg1.head()

0    27
1    27
2    26
3    30
4    22
Name: hgwy_mpg, dtype: int64

In [ ]:
mpg2.head()

121    38
176    32
24     38
88     30
72     18
Name: hgwy_mpg, dtype: int64

In [ ]:
stats.ttest_1samp(mpg2, 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [ ]:
### YOUR WORK HERE
sample2 = df.sample(100, random_state = 30)

sample2.head()

symboling norm_loss           make  ... city_mpg hgwy_mpg  price
121          1       154       plymouth  ...       31       38   6692
176         -1        65         toyota  ...       27       32  10898
24           1       148          dodge  ...       31       38   6229
88          -1       137     mitsubishi  ...       23       30   9279
72           3       142  mercedes-benz  ...       16       18  35056

[5 rows x 26 columns]

In [ ]:
mpg3 = sample2["hgwy_mpg"]

print(mpg3.shape)

mpg3.head()

(100,)


121    38
176    32
24     38
88     30
72     18
Name: hgwy_mpg, dtype: int64

In [ ]:
mpg3.mean()

30.64

In [ ]:
stats.ttest_1samp(mpg3, 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

Because we used different sample size

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [ ]:
column_headers1 = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

In [ ]:
### YOUR WORK HERE
df1 = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data", header = None, names = column_headers1)

df1.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [ ]:
df1 = df1.replace({'y': 1, 'n': 0, "?": np.NaN})

df1.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [ ]:
rep = df1[df1['party'] == 'republican']
dem = df1[df1['party'] == 'democrat']

In [ ]:
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [ ]:
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [ ]:
sample_bill = df1.sample(20, random_state=42)

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

### Null Hypothesis:

Democrats support for "budget" bill is not greater than republican support for "budget" bill.

### Alternative Hypothesis:

Democrats support for "budget" bill is greater than republican support for "budget" bill.

### Confidence level: 99%

In [ ]:
### YOUR WORK 
stats.ttest_ind(rep["budget"], dem["budget"], nan_policy= "omit")


Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

In [ ]:
dem["budget"].mean()

0.8884615384615384

In [ ]:
rep["budget"].mean()

0.13414634146341464

### Conclusion:

Based on t-statistic of -23.21 and a p-value of 2.07 to e-77, I reject the null hypothesis that democrats support for "budget" bill is not greater than republican support for "budget" bill.

### Null Hypothesis:

Republicans support for "" bill is not greater than democrats support for "" bill

### Alternative Hypothesis:

Republican support for "" bill is greater than democrats support for "" bill

### Confidence level: 99%

In [ ]:
stats.ttest_ind(dem["crime"], rep["crime"], nan_policy = "omit")

Ttest_indResult(statistic=-16.342085656197696, pvalue=9.952342705606092e-47)

In [95]:
dem['crime'].mean()

0.35019455252918286

In [97]:
rep['crime'].mean()

0.9813664596273292

### Conclusion:

Based on t-statistic of -16.34 and a p-value of 9.95 to e-47, I reject Null hypothesis that republicans support for "crime" is not greater than democrat support for "crime" bill

### Null Hypothesis:

Republican levels of support for "water-project" bill are different than democrat levels of support for "water-project" bill

### Alternative Hypothesis:

Republican support for "water-project" bill is equal to democratc levels of support for "water-project" bill

### Confidence level: 90%

In [102]:
stats.ttest_ind(dem["water-project"], rep["water-project"], nan_policy = "omit")

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

In [ ]:
dem["water-project"].mean()

0.502092050209205

In [94]:
rep["water-project"].mean()

0.5067567567567568

### Conclusion:

Based on t-statistic of -0.089 and a p-value of 0.93,I fail to reject the null hypothesis that Republican support for "water-project" bill is different than democrat levels of support for "water-project" bill

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE